src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -219,14 +249,19 @@ public IList Get(QueryKey key, ICacheAssembler[] returnTypes, bool isNaturalKeyL
 					continue;
 				}
 
+				var timestamp = GetResultsMetadata(cacheable, out var aliases); var key = keys[i];
+				if (key.ResultTransformer?.AutoDiscoverTypes == true && HasResults(cacheable))
+				{
+					key.ResultTransformer.SupplyAutoDiscoveredParameters(queryParameters[i].ResultTransformer, aliases);
+				}
+
 				var querySpaces = spaces[i];
 				if (queryParameters[i].NaturalKeyLookup || querySpaces.Count == 0)
 					continue;
 
 				spacesToCheck.Add(querySpaces);
 				checkedSpacesIndexes.Add(i);
-				// The timestamp is the first element of the cache result.
-				checkedSpacesTimestamp.Add((long) cacheable[0]);
+				checkedSpacesTimestamp.Add(timestamp);
```

Comment:
For aliases, we need to get them before short-circuiting if no timestamp checking is needed. So timestamp retrieval is also moved before the short-circuit.